contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,870 @@
+         || module.targetShare != NOR_STAKING_MODULE_TARGET_SHARE_BP
+         || module.status != uint8(StakingModuleStatus.Active)
+         || keccak256(abi.encodePacked(module.name)) != keccak256(abi.encodePacked(NOR_STAKING_MODULE_NAME))
+         || module.lastDepositAt != block.timestamp
+         || module.lastDepositBlock != block.number
+         || module.exitedValidatorsCount != 0
+        ) {
+            revert WrongStakingModuleParameters();
+        }
+    }
+
+    function _assertCorrectOracleAndConsensusContractsBinding(IBaseOracle oracle, IHashConsensus hashConsensus) internal view {
+        if (
+            oracle.getConsensusContract() != address(hashConsensus)
+         || hashConsensus.getReportProcessor() != address(oracle)
+        ) {
+            revert IncorrectOracleAndHashConsensusBinding(address(oracle), address(hashConsensus));
+        }
+    }
+
+    function _checkContractVersions() internal view {
+        _assertContractVersion(_lido, EXPECTED_FINAL_LIDO_VERSION);
+        _assertContractVersion(_nodeOperatorsRegistry, EXPECTED_FINAL_NODE_OPERATORS_REGISTRY_VERSION);
+        _assertContractVersion(_legacyOracle, EXPECTED_FINAL_LEGACY_ORACLE_VERSION);
+        _assertContractVersion(_accountingOracle, EXPECTED_FINAL_ACCOUNTING_ORACLE_VERSION);
+        _assertContractVersion(_stakingRouter, EXPECTED_FINAL_STAKING_ROUTER_VERSION);
+        _assertContractVersion(_validatorsExitBusOracle, EXPECTED_FINAL_VALIDATORS_EXIT_BUS_ORACLE_VERSION);
+        _assertContractVersion(_withdrawalQueue, EXPECTED_FINAL_WITHDRAWAL_QUEUE_VERSION);
+        _assertContractVersion(_withdrawalVault, EXPECTED_FINAL_WITHDRAWAL_VAULT_VERSION);
+    }
+
+    function _assertContractVersion(IVersioned versioned, uint256 expectedVersion) internal view {
+        if (versioned.getContractVersion() != expectedVersion) {
+            revert InvalidContractVersion(address(versioned), expectedVersion);
+        }
+    }
+
+    function _transferOZAdminFromThisToAgent(IAccessControlEnumerable accessControlled) internal {
+        accessControlled.grantRole(DEFAULT_ADMIN_ROLE, _agent);
+        accessControlled.renounceRole(DEFAULT_ADMIN_ROLE, address(this));
+    }
+
+    function _resumeWithdrawalQueue() internal {
+        IWithdrawalQueue wq = _withdrawalQueue;
+        bytes32 resume_role = wq.RESUME_ROLE();
+        wq.grantRole(resume_role, address(this));
+        wq.resume();
+        wq.renounceRole(resume_role, address(this));
+    }
+
+    // To be strict need two almost identical _resume... function because RESUME_ROLE and resume()
+    // do not actually belong to PausableUntil contract
+    function _resumeValidatorsExitBusOracle() internal {
+        IValidatorsExitBusOracle vebo = _validatorsExitBusOracle;
+        bytes32 resume_role = vebo.RESUME_ROLE();
+        vebo.grantRole(resume_role, address(this));
+        vebo.resume();
+        vebo.renounceRole(resume_role, address(this));
+    }
+
```

Comment:
It would be great to have a bare minimum of events in the template contract to check against